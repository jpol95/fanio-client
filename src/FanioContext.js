import React from 'react'

const FanioContext = React.createContext({
    fandomList: [],
    reviewList: [],
    sectionList: [], 
    subList: [],
    tagList: [],
    reviewTagList: [],
    installmentList: [],
    handleSubmitReview: () => {}, 
    handleAddFandom: () => {}, 
    handleSubmitInstallments: () => {}, 
    handleGetReview: () => {}
})

export default FanioContext