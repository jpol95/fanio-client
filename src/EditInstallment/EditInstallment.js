import React from "react";
import FanioContext from "../FanioContext";
import FetchService from "../FetchService";
import typeList from "../type-list";

export default class InstallDropDown extends React.Component {
  static contextType = FanioContext;

  types = Object.keys(typeList);

  state = {
    type: {value: "", touched: false},
    title: {value: "", touched: false},
    loaded: false
  };
  
  fandomId = Number(this.props.match.params.fandomId)
  userId = Number(this.props.match.params.userId)
  installmentId = Number(this.props.match.params.installmentId)

  //make a service class for this type of map and filter logic


  componentDidUpdate(){
      if (!this.getInstallment() || this.state.loaded) return
      this.setState({
          type: {value: this.getInstallment().type, touched: false},
          title: {value: this.getInstallment().title, touched: false}, 
          loaded: true
      })
  }

  getInstallment() {
      return this.context.installmentList.find(installment => installment.id === this.installmentId)
  }

  handleTypeChange = (e) => {
    this.setState({
      ...this.state,
      type: {value: e.target.value, touched: true},
    });
  };

  handleTitleChange = (e) => {
   this.setState({
      ...this.state,
      title: {value: e.target.value, touched: true},
    });
  };

  invalidType = () => {
    return  !this.state.type.value
  }

  invalidTitle = () => {
    return  !this.state.title.value
  }

  handleEditInstallment = async (e) => {
    e.preventDefault()
    const installCopy = {type: this.state.type.value, title: this.state.title.value, id: this.installmentId}
    const installment = await FetchService.patchInstallment(installCopy)
    this.props.history.push(`/users/${this.userId}/fandom-view/${this.fandomId}`)
    this.context.handleEditInstallment(installment)
  }

  render() {
    // console.log(this.getInstallment())
    if (!this.getInstallment()) return null
    return (
      <form onSubmit={this.handleEditInstallment}>
        <label htmlFor={`type-${this.installmentId}`}>
          What type of installment is this?
        </label>
        {this.invalidType() && <div className="error">Type is required</div>}
        <select
          onChange={this.handleTypeChange}
          id={`type-${this.installmentId}`}
        >
          <option></option>
          {this.types &&
            this.types.map((type) => {
            if (type === this.getInstallment().type) return <option value={type} selected>{type}</option>
            return <option value={type} >{type}</option>})}
        </select>
        {this.state.type !== "" && (
          <React.Fragment>
            <label>What is the name of this {this.state.type.value}?</label>
            {this.invalidTitle() && <div className="error">Title is required</div>}
            <input defaultValue={this.state.title.value} onChange={this.handleTitleChange} type="text" />
          </React.Fragment>
        )}
        <br />
        <button disabled={this.invalidType() || this.invalidTitle()} className="edit-installment-button" type="submit">Submit</button>
     </form>
    );
  }
}
