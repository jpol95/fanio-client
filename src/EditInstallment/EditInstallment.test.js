import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import EditInstallment from './EditInstallment';
import FanioContext from '../FanioContext'

describe('Edit Installment component', () => {
    const props = {
    match: {
        params: {
            userId: 1, 
            fandomId: 1, 
            installmentId: 1
        }
    }, 
    installId: 1
}
const installmentList = [
    { id: 1, title: "Doctor Who TV Series", typeId: 1, fandomId: 1 },
    { id: 2, title: "Parks and Rec TV Series", typeId: 1, fandomId: 3 },
    { id: 3, title: "Buffy TV Series", typeId: 1, fandomId: 4 },
    { id: 4, title: "Buffy Comic Series", typeId: 4, fandomId: 4 },
    { id: 5, title: "Supernatural TV Series", typeId: 1, fandomId: 2 },
    { id: 6, title: "Harry Potter Books", typeId: 2, fandomId: 5 },
  ];
  

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<FanioContext.Provider value={{installmentList: installmentList}}> <EditInstallment {...props} /> </FanioContext.Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<FanioContext.Provider value={{installmentList: installmentList}}> <EditInstallment {...props} /> </FanioContext.Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });
});