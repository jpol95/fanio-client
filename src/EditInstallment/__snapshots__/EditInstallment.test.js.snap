// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Installment component renders the UI as expected 1`] = `
Array [
  " ",
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="type-1"
    >
      What type of installment is this?
    </label>
    <div
      className="error"
    >
      Type is required
    </div>
    <select
      id="type-1"
      onChange={[Function]}
    >
      <option />
      <option
        value="Show"
      >
        Show
      </option>
      <option
        value="Book series"
      >
        Book series
      </option>
      <option
        value="Movie series"
      >
        Movie series
      </option>
      <option
        value="Comic series"
      >
        Comic series
      </option>
    </select>
    <label>
      What is the name of this 
      
      ?
    </label>
    <div
      className="error"
    >
      Title is required
    </div>
    <input
      defaultValue=""
      onChange={[Function]}
      type="text"
    />
    <br />
    <button
      className="edit-installment-button"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </form>,
  " ",
]
`;
